src/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs

```diff
@@ -219,6 +219,29 @@ private static (CompletionList CompletionList, bool IsIncomplete) FilterCompleti
             // Next, we sort the list based on the pattern matching result.
             matchResultsBuilder.Sort(MatchResult.SortingComparer);
 
+            // Determine if the list should be hard selected or soft selected.
+            bool isHardSelection;
+            if (matchResultsBuilder.Count > 0)
+            {
+                var bestResult = GetBestCompletionItemSelectionFromFilteredResults(matchResultsBuilder);
+                isHardSelection = CompletionService.IsHardSelection(bestResult.CompletionItem, bestResult.ShouldBeConsideredMatchingFilterText, completionList.SuggestionModeItem != null, filterText);
```

Comment:
is this part needed to handle soft-selecting `_`? I think we could just use the result of `IsAllPunctuation(filterText)` to decide both selection mode and isIncomplete

Reply:
Ah interesting, I can make that change